src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/NativeLayoutInfoNode.cs

```diff
@@ -0,0 +1,257 @@
+        {
+            if (_writerStream != null)
+            {
+#if DEBUG
+                // Sanity check... We should not write new items to the native layout after 
+                // we've already saved it.
+
+                MemoryStream debugStream = new MemoryStream();
+                _writer.Save(debugStream);
+                byte[] debugStreamBytes = debugStream.ToArray();
+                byte[] nativeLayoutInfoBytes = _writerStream.ToArray();
+                Debug.Assert(debugStreamBytes.Length == nativeLayoutInfoBytes.Length);
+                for (int i = 0; i < debugStreamBytes.Length; i++)
+                    Debug.Assert(debugStreamBytes[i] == nativeLayoutInfoBytes[i]);
+#endif
+                return;
+            }
+
+            _writerStream = new MemoryStream();
+            _writer.Save(_writerStream);
+        }
+
+        public override ObjectData GetData(NodeFactory factory, bool relocsOnly = false)
+        {
+            // This node does not trigger generation of other nodes.
+            if (relocsOnly)
+                return new ObjectData(Array.Empty<byte>(), Array.Empty<Relocation>(), 1, new ISymbolNode[] { this });
+
+            SaveNativeLayoutInfoWriter();
+
+            byte[] nativeLayoutInfoBytes = _writerStream.ToArray();
+
+            _endSymbol.SetSymbolOffset(nativeLayoutInfoBytes.Length);
+
+            return new ObjectData(nativeLayoutInfoBytes, Array.Empty<Relocation>(), 1, new ISymbolNode[] { this, _endSymbol });
+        }
+
+        #region Vertex Building Functions
+        public Vertex GetNativeLayoutInfoSignatureForLdToken(NodeFactory factory, MethodDesc method)
+        {
+            // TODO: option to return the uninstantiated signature info. Current implementation will only encode 
+            // the instantiated signature (i.e. with containing type, return type, args, etc... encoded as external types)
+
+            Vertex signature = null;
+
+            Vertex containingType = GetNativeLayoutInfoSignatureForEEType(factory, method.OwningType);
+            Vertex nameAndSig = _writer.GetMethodNameAndSigSignature(
+                method.Name,
+                GetNativeLayoutInfoSignatureForMethodSignature(factory, method.GetTypicalMethodDefinition()));
+            Vertex[] args = null;
+            MethodFlags flags = 0;
+
+            if (method.HasInstantiation)
+            {
+                flags |= MethodFlags.HasInstantiation;
+                args = new Vertex[method.Instantiation.Length];
+                for (int i = 0; i < args.Length; i++)
+                    args[i] = GetNativeLayoutInfoSignatureForEEType(factory, method.Instantiation[i]);
+            }
+
+            signature = _writer.GetMethodSignature((uint)flags, 0, containingType, nameAndSig, args);
+
+            return _ldTokenInfoSection.Place(signature);
+        }
+
+        public Vertex GetNativeLayoutInfoSignatureForEEType(NodeFactory factory, TypeDesc type)
```

Comment:
I'm not sure I like that these are public methods that return a `Vertex`.

Ideally, these should return some kind of wrapper (maybe even a `DependencyNodeCore` descendant?) that lets you do two things:

1. Get a cookie/offset of the `Vertex`
2. Get a list of dependencies (dependency nodes) that are recursively needed by the Vertex

The fact that 2) is not implemented means that every consumer of this API needs to have intimate knowledge of the format details (e.g. of the fact that native layout for *array of Foo* needs EEType for `Foo`, and not *array of Foo*). This will be hard to keep in sync over time (not generating enough artifacts will lead to linker errors, and generating too many will lead to size on disk bugs). A fix for this might be also relevant to the conversation in pull request #2436.

Cc @davidwrighton